libyul/optimiser/Metrics.cpp

```diff
@@ -70,8 +70,14 @@ size_t CodeWeights::costOf(Expression const& _expression) const
 		return functionCallCost;
 	else if (holds_alternative<Identifier>(_expression))
 		return identifierCost;
-	else if (holds_alternative<Literal>(_expression))
-		return literalCost;
+	else if (Literal const* literal = get_if<Literal>(&_expression))
+	{
+		// Avoid strings because they could be longer than 32 bytes.
+		if (literal->kind != LiteralKind::String && valueOfLiteral(*literal) == 0)
+			return literalZeroCost;
+		else
+			return literalCost;
+	}
```

Comment:
I think you are over-estimating the abilities of the optimizer :)
The idea here is that zero is a very common value and that it can very likely be DUPed from somewhere on the stack.
This is of course not exact, but with this assumption, the literal zero is similar to a variable reference, which also has a cost of zero.

Reply:
Unrelated question about variable refs: I feel that assigning a cost of zero to them may be sub-optimal because the fuzzer found several instances where a literal is replaced by a var decl followed by a var ref. The latter leads to additional EVM instructions, increasing gas cost.

But perhaps all of this is not exact and also not something to be concerned with since we plan to run the bytecode optimiser on the output of Yul->EVM at some point.

Reply:
Indeed, for all the optimizer steps we add a bonus if we think that this could lead to further optimizations later on by other steps.